### DISCUSSION TITLE
WebRTC and Blazor

### DISCUSSION THREAD

**Discussion by sipsorcery** [sipsorcery-org/sipsorcery#434] [General]
I'm interested in any experiences anyone has had using WebRTC in a Blazor app. I know next to nothing about Blazor. It's always on my ToDo list but Unity3D or Xamarin always seems to get in ahead of it.

With my very limited knowledge the way I understand how WebRTC and Blazor can work are:

 - Blazor Server: ~~Doesn't use the Browser's WebRTC functionality so would require a .NET WebRTC library of some kind on the server.~~ Use JS interop and the Browser's WebRTC functions.
 - Blazor WASM: Hosts the mono .NET runtime (compiled to WASM) in the browser so could use JS Interop and the Browser's WebRTC functions or could use a .NET WebRTC library.
 - Blazor Mobile: Similarly to WASM could use a .NET WebRTC library or could call out to an existing WebRTC API on the mobile OS, if available.

Is this correct?

**Comment by Terricide** [comment]
I think Blazor server/Wasm you would want to use javascript wrappers.
Blazor mobile I think it would make sense to use a .net webrtc library.

The problem with Blazor server is everything would be a roundtrip to the server and back, so for something like datachannels it makes zero sense. However I would think it would be fine to start a video call.

**Reply by sipsorcery** [reply]
So in the case of your [Blazor WebRTC example](https://github.com/Terricide/Blazor.WebRTC) you are using JS Interop and wrapping the Browser's WebRTC classes? Is the main goal of that to allow the WebRTC functions to be called in C# to avoid extra Javascript logic?

**Reply by Terricide** [reply]
Yeah the goal is to avoid writing javascript and to be able to reuse c# libraries.

I could create a library with a c# class that I can reuse in the browser and remote pc for example without having to create a duplicate javascript object.

My library could have been written more easily using blazor javascript isolation but blazor json serializes every method call and to pass binary data around I had to convert it to base64. 

I wanted the library to be high performance so I looked at how mono wrote their ClientWebsocket class

**Reply by Terricide** [reply]
I will probably end up with two versions one higher performing wasm one and one that works with blazor server since I find the debugging experience with blazor server to be better right now.

**Reply by Terricide** [reply]
Someone else got a bit farther then me with WebRTC and Blazor wasm, I only got as far as supporting DataChannels

https://github.com/melihercan/WebRTCme

**Reply by sipsorcery** [reply]
Thanks for the link. That's a fairly comprehensive implementation.

**Comment by sipsorcery** [comment]
Apparently Blazor Apps are going to be supported within MAUI apps for .NET 6. Sounds kind of like the WebView control in a Windows Form/WPF app. It was mentioned in the [.NET Rocks podcast](https://www.dotnetrocks.com/?show=1730). I didn't spot a link but I'm sure there's more info out there somewhere.

I guess the main, or one of, uses for this would be to have the same UI logic for web, desktop and mobile apps. Something that's never quite worked out in the last 20 years but probably will one day. For WebRTC it'll be interesting to see if a Blazor/WebView control hosted by MAUI supports the full WebRTC pipeline.

**Reply by Terricide** [reply]
But if you wanted a windows service to have WebRTC capability then you might be out of luck because your not a MAUI app.
